lib/compress/huf_compress.c

```diff
@@ -367,67 +367,161 @@ static U32 HUF_setMaxHeight(nodeElt* huffNode, U32 lastNonNull, U32 maxNbBits)
  *                            Must have (maxSymbolValue + 1) entries.
  * @param[in]  count          Histogram of the symbols.
  * @param[in]  maxSymbolValue Maximum symbol value.
  * @param      rankPosition   This is a scratch workspace. Must have RANK_POSITION_TABLE_SIZE entries.
  */
-static void HUF_sort(nodeElt* huffNode, const unsigned* count, U32 maxSymbolValue, rankPos* rankPosition)
-{
+static void HUF_sort(nodeElt huffNode[], const unsigned count[], U32 const maxSymbolValue, rankPos rankPosition[]) {
     int n;
     int const maxSymbolValue1 = (int)maxSymbolValue + 1;
 
     /* Compute base and set curr to base.
-     * For symbol s let lowerRank = BIT_highbit32(count[n]+1) and rank = lowerRank + 1.
-     * Then 2^lowerRank <= count[n]+1 <= 2^rank.
+     * For symbol s let lowerRank = HUF_getIndex(count[n]) and rank = lowerRank + 1.
+     * See HUF_getIndex to see bucketing strategy.
      * We attribute each symbol to lowerRank's base value, because we want to know where
      * each rank begins in the output, so for rank R we want to count ranks R+1 and above.
      */
     ZSTD_memset(rankPosition, 0, sizeof(*rankPosition) * RANK_POSITION_TABLE_SIZE);
     for (n = 0; n < maxSymbolValue1; ++n) {
-        U32 lowerRank = BIT_highbit32(count[n] + 1);
+        U32 lowerRank = HUF_getIndex(count[n]);
         rankPosition[lowerRank].base++;
     }
+
     assert(rankPosition[RANK_POSITION_TABLE_SIZE - 1].base == 0);
+    /* Set up the rankPosition table */
     for (n = RANK_POSITION_TABLE_SIZE - 1; n > 0; --n) {
         rankPosition[n-1].base += rankPosition[n].base;
         rankPosition[n-1].curr = rankPosition[n-1].base;
     }
-    /* Sort */
+
+    /* Insert each symbol into their appropriate bucket, setting up rankPosition table. */
     for (n = 0; n < maxSymbolValue1; ++n) {
         U32 const c = count[n];
-        U32 const r = BIT_highbit32(c+1) + 1;
-        U32 pos = rankPosition[r].curr++;
-        /* Insert into the correct position in the rank.
-         * We have at most 256 symbols, so this insertion should be fine.
-         */
-        while ((pos > rankPosition[r].base) && (c > huffNode[pos-1].count)) {
-            huffNode[pos] = huffNode[pos-1];
-            pos--;
-        }
+        U32 const r = HUF_getIndex(c) + 1;
+        U32 const pos = rankPosition[r].curr++;
         huffNode[pos].count = c;
         huffNode[pos].byte  = (BYTE)n;
     }
-}
 
+    /* Sort each bucket. */
+    for (n = RANK_POSITION_DISTINCT_COUNT_CUTOFF; n < RANK_POSITION_TABLE_SIZE - 1; ++n) {
+        U32 const bucketSize = rankPosition[n].curr-rankPosition[n].base;
+		U32 const bucketStartIdx = rankPosition[n].base;
+        if (bucketSize <= 1) {
+            continue;
+        } else if (bucketSize <= 128) {
+            HUF_insertionSort(huffNode + bucketStartIdx, bucketSize);
+        } else {
+            HUF_simpleQuickSort(huffNode + bucketStartIdx, 0, bucketSize-1);
+        }
```

Comment:
I'd recommend just using a single sort function. Generally, fast quick sorts will look at array size, and for small array sizes will fall back to insertion sort. So the quick sort could detect arrays of < K (K ~5 maybe) elements and fall back to insertion sort.

Reply:
So I think it's a net positive change, but it's hard to tell, really. 6 elements experimentally seems like an okay cutoff point. With using insertion sort within quicksort for calls of less than 6 elements, we see the following:

|     Benchmark     | Config  |   Dataset   | Ratio (5accf234) | Ratio (cfeccb43) | Ratio (cfeccb43 - 5accf234) | Speed MB/s (5accf234) | Speed MB/s (cfeccb43) | Speed MB/s (cfeccb43 - 5accf234) |
|-------------------|---------|-------------|------------------|------------------|-----------------------------|-----------------------|-----------------------|----------------------------------|
| compress          | level_1 | enwik7      | 2.43             | 2.43             | -0.0%                       | 260.90                | 259.36                | -0.6%                            |
| compress          | level_1 | enwik7_1k   | 1.75             | 1.75             | 0.0%                        | 115.55                | 117.16                | 1.4%                             |
| compress          | level_1 | silesia     | 2.88             | 2.88             | 0.0%                        | 349.45                | 347.77                | -0.5%                            |
| compress          | level_1 | silesia_16k | 2.62             | 2.62             | -0.0%                       | 247.51                | 249.12                | 0.7%                             |
| compress          | level_1 | silesia_4k  | 2.31             | 2.31             | 0.0%                        | 204.78                | 205.35                | 0.3%                             |
| compress          | level_3 | enwik7      | 2.79             | 2.79             | 0.0%                        | 144.61                | 142.33                | -1.6%                            |
| compress          | level_3 | enwik7_1k   | 1.75             | 1.75             | -0.0%                       | 87.10                 | 87.85                 | 0.9%                             |
| compress          | level_3 | silesia     | 3.18             | 3.18             | 0.0%                        | 190.72                | 187.68                | -1.6%                            |
| compress          | level_3 | silesia_16k | 2.67             | 2.67             | 0.0%                        | 182.81                | 183.10                | 0.2%                             |
| compress          | level_3 | silesia_4k  | 2.35             | 2.35             | -0.0%                       | 154.32                | 154.96                | 0.4%                             |
| compress          | level_7 | enwik7      | 3.05             | 3.05             | 0.0%                        | 66.42                 | 66.32                 | -0.1%                            |
| compress          | level_7 | enwik7_1k   | 1.79             | 1.79             | -0.0%                       | 52.59                 | 53.06                 | 0.9%                             |
| compress          | level_7 | silesia     | 3.45             | 3.45             | -0.0%                       | 87.01                 | 86.72                 | -0.3%                            |
| compress          | level_7 | silesia_16k | 2.80             | 2.80             | -0.0%                       | 51.72                 | 52.20                 | 0.9%                             |
| compress          | level_7 | silesia_4k  | 2.42             | 2.42             | 0.0%                        | 62.42                 | 62.49                 | 0.1%                             |
| compress_literals | level_1 | enwik7      | 1.49             | 1.49             | -0.0%                       | 783.73                | 778.03                | -0.7%                            |
| compress_literals | level_1 | enwik7_1k   | 1.46             | 1.46             | 0.0%                        | 264.67                | 271.91                | 2.7%                             |
| compress_literals | level_1 | silesia     | 1.30             | 1.30             | 0.0%                        | 604.35                | 604.73                | 0.1%                             |
| compress_literals | level_1 | silesia_16k | 1.29             | 1.29             | 0.0%                        | 558.75                | 557.81                | -0.2%                            |
| compress_literals | level_1 | silesia_1k  | 1.28             | 1.28             | 0.0%                        | 265.02                | 272.08                | 2.7%                             |
| compress_literals | level_1 | silesia_4k  | 1.29             | 1.29             | 0.0%                        | 276.64                | 283.87                | 2.6%                             |
| compress_literals | level_3 | enwik7      | 1.34             | 1.34             | 0.0%                        | 640.60                | 630.37                | -1.6%                            |
| compress_literals | level_3 | enwik7_1k   | 1.40             | 1.40             | -0.0%                       | 201.50                | 205.74                | 2.1%                             |
| compress_literals | level_3 | silesia     | 1.17             | 1.17             | -0.0%                       | 499.56                | 501.60                | 0.4%                             |
| compress_literals | level_3 | silesia_16k | 1.21             | 1.21             | 0.0%                        | 492.30                | 492.93                | 0.1%                             |
| compress_literals | level_3 | silesia_1k  | 1.23             | 1.23             | 0.0%                        | 223.39                | 228.90                | 2.5%                             |
| compress_literals | level_3 | silesia_4k  | 1.24             | 1.24             | 0.0%                        | 239.28                | 244.97                | 2.4%                             |
| compress_literals | level_7 | enwik7      | 1.29             | 1.29             | 0.0%                        | 567.40                | 557.35                | -1.8%                            |
| compress_literals | level_7 | enwik7_1k   | 1.39             | 1.39             | -0.0%                       | 189.78                | 193.95                | 2.2%                             |
| compress_literals | level_7 | silesia     | 1.15             | 1.15             | -0.0%                       | 461.80                | 464.98                | 0.7%                             |
| compress_literals | level_7 | silesia_16k | 1.21             | 1.21             | -0.0%                       | 477.96                | 478.90                | 0.2%                             |
| compress_literals | level_7 | silesia_1k  | 1.22             | 1.22             | 0.0%                        | 215.32                | 220.73                | 2.5%                             |
| compress_literals | level_7 | silesia_4k  | 1.23             | 1.23             | 0.0%                        | 232.33                | 237.94                | 2.4%                             |


The main issue just seems to be that e2e compression on normal sized blocks seems to be consistently slower by a bit. It doesn't really make sense that that should happen since literals compression is typically still faster by a bit, so that might just be mostly noise.

Reply:
Update: 32 elements actually seems like a better breakpoint. I've compared 8, 16, and 32 element thresholds. You can see the raw data here: https://pastebin.com/raw/jWd7Gu8R

I will try and see if some of the insertion sort optimizations in glibc's `qsort()` provide any additional benefit